hw05-AOP/src/main/java/ru/otus/proxy/IOC.java

```diff
@@ -0,0 +1,63 @@
+package ru.otus.proxy;
+
+import ru.otus.annotations.Log;
+import ru.otus.domain.Constants;
+import ru.otus.utils.ReflectionUtils;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public abstract class IOC {
+
+	private static final Set<String> methodsForLogging = new HashSet<>();
+
+	@SuppressWarnings("unchecked")
+	public static <T> T getInstance(final T obj, final Class<T> implInterface) {
+		final Method[] methods = obj.getClass().getMethods();
+		final Set<Method> annotatedMethods = ReflectionUtils.filterMethodsByAnnotation(methods, Log.class);
+
+		if (annotatedMethods.isEmpty()) {
+			throw new IllegalArgumentException("Not found annotation: " + Log.class.getCanonicalName());
+		}
+
+		final Set<String> namesAndParams = annotatedMethods.stream().map(ReflectionUtils::extractNameAndParams).collect(Collectors.toSet());
+
+		methodsForLogging.addAll(namesAndParams);
+
+		return (T) Proxy.newProxyInstance(IOC.class.getClassLoader(), new Class<?>[]{implInterface}, new LogInvocationHandler<>(obj));
```

Comment:
```suggestion
		return (T) Proxy.newProxyInstance(IOC.class.getClassLoader(), obj.getClass().getInterfaces(), new LogInvocationHandler<>(obj));
```

Можно убрать параметр `implInterface`